### ISSUE TITLE
1.4-beta.1 - fourseven:scss package broken

### ISSUE THREAD

**Issue by mikestokes** [meteor/meteor#7297]
Running OS X El Capitan and upgrading to 1.4-beta.1 there is an issue that seems to be related to the npm and/or node update with the fourseven:scss package for compiling sass files.

Re-creation:

1. Upgrade to Meteor 1.4-beta.1
2. Install the fourseven:scss package using the "meteor add" command
3. Remove NPM packages folder: rm -rf node_modules
4. Install NPM packages: meteor npm install
5. Run meteor

`[[[[[ ~/git/fletchers/fd.shopfloor.meteor ]]]]]

=> Started proxy.
=> Started MongoDB.
=> Errors prevented startup:

   While loading plugin `compileScssBatch` from package `fourseven:scss`:

   /Users/mike/.meteor/packages/fourseven_scss/.3.4.3.fcqret++os.osx.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.osx.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/lib/extensions.js:158:11:
   The `libsass` binding was not found in
   /Users/mike/.meteor/packages/fourseven_scss/.3.4.3.fcqret++os.osx.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.osx.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/vendor/darwin-x64-46/binding.node
   This usually happens because your node version has changed.
   Run `npm rebuild node-sass` to build the binding for your current node
   version.
   at Object.sass.getBinaryPath
   (/Users/mike/.meteor/packages/fourseven_scss/.3.4.3.fcqret++os.osx.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.osx.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/lib/extensions.js:158:11)
   at Object.<anonymous>
   (/Users/mike/.meteor/packages/fourseven_scss/.3.4.3.fcqret++os.osx.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.osx.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/lib/index.js:16:36)
   at Module._compile (module.js:409:26)
   at Object.Module._extensions..js (module.js:416:10)
   at Module.load (module.js:343:32)
   at Module.Mp.load
   (/Users/mike/.meteor/packages/meteor-tool/.1.4.0-beta.1.1jq88fv++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/reify/node/runtime.js:16:23)
   at Function.Module._load (module.js:300:12)
   at Module.require (module.js:353:17)
   at require (internal/module.js:12:17)
   at Object.<anonymous> (/tools/isobuild/bundler.js:1570:22)
   at Object.require (/tools/isobuild/bundler.js:1547:11)
   at meteorInstall.node_modules.meteor.compileScssBatch.plugin.compile-scss.js
   (packages/compileScssBatch/plugin/compile-scss.js:3:18)
   at fileEvaluate (packages/modules-runtime.js:158:9)
   at require (packages/modules-runtime.js:92:16)
   at <runJavaScript-29>:349:1
   at <runJavaScript-29>:355:3

=> Your application has errors. Waiting for file change.`

There is an [open issue in the forseven:scss repo](https://github.com/fourseven/meteor-scss/issues/184) but I think this may point to a Meteor bug.

**Comment by clayne11** [comment]
I would recommend using [`nathantreid:css-modules `](https://github.com/nathantreid/meteor-css-modules) instead. It has built in support for PostCSS and generally more configurable. You don't have to use CSS modules to use it, you can use and PostCSS plugins that you'd like and it also compiles SCSS.

**Comment by mikestokes** [comment]
Thanks @clayne11 I'll give it a try for SCSS.

**Comment by meteorza** [comment]
I also have 

I use poetic:materialize-scss which auto installs fourseven:scss, so I can not use an alternative scss compiler. I get the same error as mikestokes when using meteor 1.4 beta.

**Comment by nathantreid** [comment]
Referring to [the Meteor 1.4 announcement](https://forums.meteor.com/t/coming-soon-meteor-1-4/23260/1), my guess is that you need to install a compiler toolchain in order to build libsass (since it's still in beta, I don't know if this documented yet, and unfortunately I don't where you'd get this toolchain):

> To solve this problem for Node 4 and any future Node updates, we’re going to require Meteor users who rely on packages with binary dependencies to install a compiler toolchain on their development and build machines. We’ll make sure this is a straightforward and well documented process. In addition, we’ll ship Meteor’s core dependencies with pre-built binaries. This will mean that new and simple app development will continue to work without requiring a compiler toolchain. 

My css-modules package works because I require the user to `meteor npm install node-sass` instead of including it in the package (which also allows the user to pick the version of node-sass that they want to use).

**Comment by ramijarrar** [comment]
@nathantreid "compiler toolchain" would refer to the requirements for node-gyp (i.e python & C/C++ compiler). But node-sass ships with pre-built binaries so I don't think that is the issue.

I think your approach of allowing people to use their own version of node-sass is definitely better (and loading npm dependencies in packages can lead to duplication) but this should still work regardless.

**Comment by ramijarrar** [comment]
@sebakerckhof Would it be possible to release a new version of ```fourseven:scss``` which loads node-sass as a "peer dependency" (using [tmeasday:check-npm-versions](https://github.com/tmeasday/check-npm-versions)) ?

**Comment by donflopez** [comment]
I've solved it cloning the fourseven:scss package into my local packages folder and updating the node-sass version in the package.js file to 3.8.0

**Comment by sebakerckhof** [comment]
@ramijarrar thanks for looking into this. 
TBH, I didn't really know what peer dependencies were. I looked it up: https://nodejs.org/en/blog/npm/peer-dependencies/ and I don't find the description fitting for this use case. So I'll first see if upgrading node-sass to 3.8.0 fixes it for everyone and otherwise I'd rather wait for input from @benjamn

**Comment by ramijarrar** [comment]
If you read my comments above I was referring to [this](https://guide.meteor.com/writing-atmosphere-packages.html#peer-npm-dependencies) approach for loading npm dependencies in an atmosphere package.

**Comment by ramijarrar** [comment]
Found some more information about this issue in the node-sass changelog (see [v3.5.1](https://github.com/sass/node-sass/releases/tag/v3.5.1)), turns out Node 4.x was not supported on 64bit OSX until v3.7.0 so upgrading as suggested by @donflopez should get rid of that for everyone.

**Comment by sebakerckhof** [comment]
3.7.0 adds node 6 support.
If you look here: https://github.com/sass/node-sass/releases/tag/v3.4.2 it states OSX 64-bit was already supported with node 4.

3.5.1 just states there are better error messages and it takes osx 64-bit as an example.

Anyway, will publish a new version later today.

**Comment by ramijarrar** [comment]
@sebakerckhof Looks like I misinterpreted that :) 

I'll try again with the new release as soon as that is published. Are you still just planning to update node-sass?

**Comment by sebakerckhof** [comment]
I just released fourseven:scss 3.8.0, with node-sass 3.8.0.

Tested with 1.4-beta.2, error is still present, although a somewhat more useful error message:
```
   While loading plugin `compileScssBatch` from package `fourseven:scss`:

   C:\Users\seke\AppData\Local\.meteor\packages-from-server\kndedu01.barco.comX3005\fourseven_scss\3.8.0\plugin.compileScssBatch.os.windows.x86_32\npm\node_modules\meteor\compileScssBatch\node_modules\node-sass\lib\index.js:15:11:
   Missing binding
   C:\Users\seke\AppData\Local\.meteor\packages-from-server\kndedu01.barco.comX3005\fourseven_scss\3.8.0\plugin.compileScssBatch.os.windows.x86_32\npm\node_modules\meteor\compileScssBatch\node_modules\node-sass\vendor\win32-ia32-46\binding.node
   Node Sass could not find a binding for your current environment: Windows 32-bit with Node.js 4.x

   Found bindings for the following environments:
   - Windows 32-bit with Node 0.10.x

   This usually happens because your environment has changed since running `npm install`.
   Run `npm rebuild node-sass` to build the binding for your current environment.
   at Object.<anonymous>
   (C:\Users\seke\AppData\Local\.meteor\packages-from-server\kndedu01.barco.comX3005\fourseven_scss\3.8.0\plugin.compileScssBatch.os.windows.x86_32\npm\node_modules\meteor\compileScssBatch\node_modules\node-sass\lib\index.js:15:11)
   at Module._compile (module.js:409:26)
   at Object.Module._extensions..js (module.js:416:10)
   at Module.load (module.js:343:32)
   at Module.Mp.load
   (C:\Users\seke\AppData\Local\.meteor\packages-from-server\kndedu01.barco.comX3005\meteor-tool\1.4.0-beta.2\mt-os.windows.x86_32\dev_bundle\lib\node_modules\reify\node\runtime.js:16:23)
   at Function.Module._load (module.js:300:12)
   at Module.require (module.js:353:17)
   at require (internal/module.js:12:17)
   at Object.<anonymous> (C:\tools\isobuild\bundler.js:1570:22)
   at Object.require (C:\tools\isobuild\bundler.js:1547:11)
   at meteorInstall.node_modules.meteor.compileScssBatch.plugin.compile-scss.js (packages/compileScssBatch/plugin/compile-scss.js:3:18)
   at fileEvaluate (packages/modules-runtime/.npm/package/node_modules/install/install.js:153:1)
   at require (packages/modules-runtime/.npm/package/node_modules/install/install.js:82:1)
   at <runJavaScript-41>:349:1
   at <runJavaScript-41>:355:3
```

This makes sense, since it was build on 0.10.x (meteor build server). But Meteor should have ran the rebuild which it didn't I suppose?

Will wait for input from @benjamn

**Comment by sebakerckhof** [comment]
Btw @benjamn : OSX build servers don't seem to work. It takes a very long time to connect then it asks me for login and just exits.

**Comment by mikeifomin** [comment]
Update to Meteor 1.4-beta.7  does not help

**Comment by paralin** [comment]
Any updates here?

**Comment by StorytellerCZ** [comment]
Just updated to `1.4-beta.9` and still an issue. :cry:

**Comment by abernix** [comment]
To those that are still experiencing this issue, what OS are you using?

**Comment by mikeifomin** [comment]
I have the same in 1.4-beta.9 on OS X and Ubuntu 14.04.

#### Reproduce

```
   meteor create test-app --release=1.4-beta.9 && cd test-app && meteor add fourseven:scss && meteor run
```

#### Output on Ubuntu
```
=> Started proxy.
=> Started MongoDB.
=> Errors prevented startup:

   While loading plugin `compileScssBatch` from package `fourseven:scss`:

   /root/.meteor/packages/fourseven_scss/.3.8.0_1.1vehobq++os.linux.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.linux.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/lib/index.js:15:11:
   Missing binding
   /root/.meteor/packages/fourseven_scss/.3.8.0_1.1vehobq++os.linux.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.linux.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/vendor/linux-x64-46/binding.node
   Node Sass could not find a binding for your current environment: Linux 64-bit with Node.js 4.x

   Found bindings for the following environments:
   - Unknown environment (.meteor-portable)
   - Linux 64-bit with Node 0.10.x

   This usually happens because your environment has changed since running `npm install`.
   Run `npm rebuild node-sass` to build the binding for your current environment.
   at Object.<anonymous>
   (/root/.meteor/packages/fourseven_scss/.3.8.0_1.1vehobq++os.linux.x86_64+web.browser+web.cordova/plugin.compileScssBatch.os.linux.x86_64/npm/node_modules/meteor/compileScssBatch/node_modules/node-sass/lib/index.js:15:11)
   at Module._compile (module.js:409:26)
   at Object.Module._extensions..js (module.js:416:10)
   at Module.load (module.js:343:32)
   at Module.Mp.load
   (/root/.meteor/packages/meteor-tool/.1.4.0-beta.9.1yjtbnw++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/reify/node/runtime.js:16:23)
   at Function.Module._load (module.js:300:12)
   at Module.require (module.js:353:17)
   at require (internal/module.js:12:17)
   at Object.<anonymous> (/tools/isobuild/bundler.js:1565:22)
   at Object.require (/tools/isobuild/bundler.js:1542:11)
   at meteorInstall.node_modules.meteor.compileScssBatch.plugin.compile-scss.js (packages/compileScssBatch/plugin/compile-scss.js:3:18)
   at fileEvaluate (packages/modules-runtime/.npm/package/node_modules/install/install.js:153:1)
   at require (packages/modules-runtime/.npm/package/node_modules/install/install.js:82:1)
   at <runJavaScript-41>:349:1
   at <runJavaScript-41>:355:3
```


@benjamn I think this issue should be opened.

**Comment by StorytellerCZ** [comment]
Seems to be fixed for me in `beta.10`.

**Comment by jbbr** [comment]
Works for me in `1.4-beta.10` - thanks!
I had to remove `$HOME/.meteor` and/or `.meteor/local` to get finally rid of this error message.

**Comment by finchalyzer** [comment]
Issue gone with `1.4-beta.10` on OS X 10.11.5. Only after clearing `$HOME/.meteor`

**Comment by ixdi** [comment]
I'm still experiencing this issue with `1.4-beta.13`. I tried unsuccessfully to remove `$HOME/.meteor` and `.meteor/local`.

I'm running on Arch linux with latest updates (kernel 4.6.3)

**Comment by sebakerckhof** [comment]
Works for me on Win 10

**Comment by lope** [comment]
I had some other problems upgrading cleanly from 1.3.4.1>1.3.4.3>1.3.4.4>1.4 beta 9. Abernix is going to try reproduce.
So it's hard to say what was going on with my setup with certainty because my proj/.meteor/release file was stuck on `1.3.4.3` but whatever the case is. When I tried to force `meteor run --release 1.4-beta.9` I got a similar issue.
But then I just nuked my whole ~/.meteor, manually edited proj/.meteor/release and deleted proj/.meteor/dev_bundle and now using 1.4-beta-14 without issues.
Linux 4.4 kernel.

**Comment by mikeifomin** [comment]
@lope thank you!

**Comment by Metrophobe** [comment]
any solution guys ? I am using the materialize:materialize build and it seems like the fourseven:scss is breaking everything. I tried the above but still am getting the same error.

**Comment by sebakerckhof** [comment]
@Metrophobe can you create a minimal reproduction?

**Comment by lope** [comment]
@mikeifomin you're welcome.
Okay guys. My build times are pretty decent since I moved all of my .jsx .scss etc files into /imports/client/ Around 2.5 seconds or less when changing client files.

I've not ventured into SSR terrirory yet, which would raise build times back into the 5.7 second territory for me. But apparently there's a hack where you can just include all the client files for SSR in ONE file and just comment out that file when you're developing for faster build times (so you won't have SSR while developing) then uncomment it for production. I really know nothing about SSR at this point, so this is just heresy.

I'm going to unsubscribe from this thread now due to too many github emails. Write @lope if you want me to see your msg if you've got something useful to share about build times and SSR.

**Comment by felixb101** [comment]
Still get that version 3.8.1 is not compatible with Meteor , any updates here?

**Comment by Metrophobe** [comment]
@felixb101 To resolve the issue and I have re-iterated this numerous times and then I got nagged by someone on the threads so I refrained from commenting was ....

if fourseveLscss  is your offending package ... goto packages.json in the .meteor folder and insert it PRIOR to the package being downloading and crashing and you're good to go!

so for example if  materialize:materialize - has a dependency fourseven:scss and you get the long crashing message.  then simply place fourseven:scss before materialize:materialize (in packages.json) and you're good to go! There is a defect in the toolchain which I had pointed out quite some time ago and its not managing all dependencies correctly but is loading them ok. 

This applies to any other error you get in any other dependency I got one in bcrypt and coffescript last time. Used this method and resolved all issues.

Hope this helps you sir.
Good Day
